sklearn/ensemble/_hist_gradient_boosting/_loss.pyx

```diff
@@ -27,9 +27,51 @@ def _update_gradients_least_squares(
 
     n_samples = raw_predictions.shape[0]
     for i in prange(n_samples, schedule='static', nogil=True):
+        # Note: a more correct exp is 2 * (raw_predictions - y_true)
+        # but since we use 1 for the constant hessian value (and not 2) this
+        # is strictly equivalent for the leaves values.
         gradients[i] = raw_predictions[i] - y_true[i]
 
 
+def _update_gradients_hessians_least_squares(
+        G_H_DTYPE_C [::1] gradients,  # OUT
+        G_H_DTYPE_C [::1] hessians,  # OUT
+        const Y_DTYPE_C [::1] y_true,  # IN
+        const Y_DTYPE_C [::1] raw_predictions,  # IN
+        const Y_DTYPE_C [::1] sample_weight):  # IN
+
+    cdef:
+        int n_samples
+        int i
+
+    n_samples = raw_predictions.shape[0]
+    for i in prange(n_samples, schedule='static', nogil=True):
+        # Note: a more correct exp is 2 * (raw_predictions - y_true) * sample_weight
+        # but since we use 1 for the constant hessian value (and not 2) this
+        # is strictly equivalent for the leaves values.
+        gradients[i] = (raw_predictions[i] - y_true[i]) * sample_weight[i]
+        hessians[i] = sample_weight[i]
+
+
+def _update_gradients_hessians_least_absolute_deviation(
+        G_H_DTYPE_C [::1] gradients,  # OUT
+        G_H_DTYPE_C [::1] hessians,  # OUT
+        const Y_DTYPE_C [::1] y_true,  # IN
+        const Y_DTYPE_C [::1] raw_predictions,  # IN
+        const Y_DTYPE_C [::1] sample_weight):  # IN
+
+    cdef:
+        int n_samples
+        int i
+
+    n_samples = raw_predictions.shape[0]
+    for i in prange(n_samples, schedule='static', nogil=True):
+        # gradient = sign(raw_predicition - y_pred) * sample_weight
+        gradients[i] = sample_weight[i] * (2 *
+                        (y_true[i] - raw_predictions[i] < 0) - 1)
+        hessians[i] = sample_weight[i]
```

Comment:
> Does this works because sample_weight is non-negative?

No, this works because:

- accounting for SW means we need to multiply gradients and hessians by SW
- without SW, the hessian of this loss is constant and is equal to 1.

Reply:
I was thinking about the math. lightgbm does the same thing with the l1 loss.

https://github.com/microsoft/LightGBM/blob/4adb9ff71f41f6b5c7a51f667a8fb9adf38cf602/src/objective/regression_objective.hpp#L225-L229

When I see derivative of `sign(x)`, I think of the dirac delta function: https://en.wikipedia.org/wiki/Sign_function

Reply:
Ah I think I see what you did here: https://github.com/scikit-learn/scikit-learn/pull/13896#discussion_r293897646

Reply:
yeah, also, if you look at the sklearn wrapper in lightgbm, all hessians and gradients are simply multiplied by SW.